File: applications/utils/import/srtm2wayinfo/srtm.py
#!/usr/bin/env python

"""Load and process SRTM data."""

from __future__ import with_statement
import ftplib
import re
import pickle
import os.path
import os
import zipfile
import array
import math

class NoSuchTileError(Exception):
    """Raised when there is no tile for a region."""
    def __init__(self, lat, lon):
        self.lat = lat
        self.lon = lon

    def __str__(self):
        return "No SRTM tile for %d, %d available!" % (self.lat, self.lon)


class WrongTileError(Exception):
    """Raised when the value of a pixel outside the tile area is requested."""
    def __init__(self, tile_lat, tile_lon, req_lat, req_lon):
        self.tile_lat = tile_lat
        self.tile_lon = tile_lon
        self.req_lat = req_lat
        self.req_lon = req_lon

    def __str__(self):
        return "SRTM tile for %d, %d does not contain data for %d, %d!" % (
            self.tile_lat, self.tile_lon, self.req_lat, self.req_lon)

class InvalidTileError(Exception):
    """Raised when the SRTM tile file contains invalid data."""
    def __init__(self, lat, lon):
        self.lat = lat
        self.lon = lon

    def __str__(self):
        return "SRTM tile for %d, %d is invalid!" % (self.lat, self.lon)

class SRTMDownloader:
    """Automatically download SRTM tiles."""
    def __init__(self, server="e0srp01u.ecs.nasa.gov",
            directory="/srtm/version2/SRTM3", cachedir="cache"):
        self.server = server
        self.directory = directory
        self.cachedir = cachedir
        if not os.path.exists(cachedir):
            os.mkdir(cachedir)
        self.filelist = {}
        self.filename_regex = re.compile(r"([NS])(\d{2})([EW])(\d{3})\.hgt\.zip")
        self.filelist_file = self.cachedir + "/filelist"
        self.ftpfile = None
        self.ftp_bytes_transfered = 0

    def loadFileList(self):
        """Load a previously created file list or create a new one if none is
            available."""
        try:
            data = open(self.filelist_file, 'rb')
        except IOError:
            print "No cached file list. Creating new one!"
            self.createFileList()
            return
        try:
            self.filelist = pickle.load(data)
        except:
            print "Unknown error loading cached file list. Creating new one!"
            self.createFileList()

    def createFileList(self):
        """SRTM data is split into different directories, get a list of all of
            them and create a dictionary for easy lookup."""
        ftp = ftplib.FTP(self.server)
        try:
            ftp.login()
            ftp.cwd(self.directory)
            continents = ftp.nlst()
            for continent in continents:
                print "Downloading file list for", continent
                ftp.cwd(self.directory+"/"+continent)
                files = ftp.nlst()
                for filename in files:
                    self.filelist[self.parseFilename(filename)] = (continent, filename)
        finally:
            ftp.close()
        # Add meta info
        self.filelist["server"] = self.server
        self.filelist["directory"] = self.directory
        with open(self.filelist_file , 'wb') as output:
            pickle.dump(self.filelist, output)

    def parseFilename(self, filename):
        """Get lat/lon values from filename."""
        match = self.filename_regex.match(filename)
        if match is None:
            # TODO?: Raise exception?
            print "Filename", filename, "unrecognized!"
            return None
        lat = int(match.group(2))
        lon = int(match.group(4))
        if match.group(1) == "S":
            lat = -lat
        if match.group(3) == "W":
            lon = -lon
        return lat, lon

    def getTile(self, lat, lon):
        """Get a SRTM tile object. This function can return either an SRTM1 or
            SRTM3 object depending on what is available, however currently it
            only returns SRTM3 objects."""
        try:
            continent, filename = self.filelist[(int(lat), int(lon))]
        except KeyError:
            raise NoSuchTileError(lat, lon)
        if not os.path.exists(self.cachedir + "/" + filename):
            self.downloadTile(continent, filename)
        # TODO: Currently we create a new tile object each time.
        # Caching is required for improved performance.
        return SRTMTile(self.cachedir + "/" + filename, int(lat), int(lon))

    def downloadTile(self, continent, filename):
        """Download a tile from NASA's server and store it in the cache."""
        ftp = ftplib.FTP(self.server)
        try:
            ftp.login()
            ftp.cwd(self.directory+"/"+continent)
            # WARNING: This is not thread safe
            self.ftpfile = open(self.cachedir + "/" + filename, 'wb')
            self.ftp_bytes_transfered = 0
            print ""
            try:
                ftp.retrbinary("RETR "+filename, self.ftpCallback)
            finally:
                self.ftpfile.close()
                self.ftpfile = None
        finally:
            ftp.close()

    def ftpCallback(self, data):
        """Called by ftplib when some bytes have been received."""
        self.ftpfile.write(data)
        self.ftp_bytes_transfered += len(data)
        print "\r%d bytes transfered" % self.ftp_bytes_transfered,


class SRTMTile:
    """Base class for all SRTM tiles.
        Each SRTM tile is size x size pixels big and contains
        data for the area from (lat, lon) to (lat+1, lon+1) inclusive.
        This means there is a 1 pixel overlap between tiles. This makes it
        easier for as to interpolate the value, because for every point we
        only have to look at a single tile.
        """
    def __init__(self, f, lat, lon):
        zipf = zipfile.ZipFile(f, 'r')
        names = zipf.namelist()
        if len(names) != 1:
            raise InvalidTileError(lat, lon)
        data = zipf.read(names[0])
        self.size = int(math.sqrt(len(data)/2)) #2 bytes per sample
        if self.size != 1201: #Currently only SRTM3 is supported
            raise InvalidTileError(lat, lon)
        self.data = array.array('h', data)
        self.data.byteswap()
        if len(self.data) != self.size * self.size:
            raise InvalidTileError(lat, lon)
        self.lat = lat
        self.lon = lon

    def _avg(self, value1, value2, weight):
        """Returns the weighted average of two values and handles the case where
            one value is None. If both values are None, None is returned.
        """
        if value1 is None:
            return value2
        if value2 is None:
            return value1
        return value2 * weight + value1 * (1 - weight)

    def calcOffset(self, x, y):
        # Datalayout
        # X = longitude
        # Y = latitude
        # Sample for size 1201x1201
        #  (   0/1200)     (   1/1200)  ...    (1199/1200)    (1200/1200)
        #  (   0/1199)     (   1/1199)  ...    (1199/1199)    (1200/1199)
        #       ...            ...                 ...             ...
        #  (   0/   1)     (   1/   1)  ...    (1199/   1)    (1200/   1)
        #  (   0/   0)     (   1/   0)  ...    (1199/   0)    (1200/   0)
        #  Some offsets:
        #  (0/1200)     0
        #  (1200/1200)  1200
        #  (0/1199)     1201
        #  (1200/1199)  2401
        #  (0/0)        1201*1200
        #  (1200/0)     1201*1201-1
        return x + self.size * (self.size - y - 1)

    def getPixelValue(self, x, y):
        #print "pixel", x, y
        assert x<self.size, "x: %d<%d" % (x, self.size)
        assert y<self.size, "y: %d<%d" % (y, self.size)
        #Same as calcOffset, inlined for performance reasons
        offset = x + self.size * (self.size - y - 1)
        value = self.data[offset]
        #print value
        if value == -32768:
            return None #-32768 is a special value for areas with no data
        return value
        

    def getAltitudeFromLatLon(self, lat, lon):
        """Get the altitude of a lat lon pair, using the four neighbouring
            pixels for interpolation.
        """
        #print "-----\nFromLatLon", lon, lat
        lat -= self.lat
        lon -= self.lon
        #print "lon, lat", lon, lat
        if lat < 0.0 or lat >= 1.0 or lon < 0.0 or lon >= 1.0:
            raise WrongTileError(self.lat, self.lon, self.lat+lat, self.lon+lon)
        x = lon * (self.size - 1)
        y = lat * (self.size - 1)
        #print "x,y", x, y
        x_int = int(x)
        x_frac = x - int(x)
        y_int = int(y)
        y_frac = y - int(y)
        #print "frac", x_int, x_frac, y_int, y_frac
        value00 = self.getPixelValue(x_int, y_int)
        value10 = self.getPixelValue(x_int+1, y_int)
        value01 = self.getPixelValue(x_int, y_int+1)
        value11 = self.getPixelValue(x_int+1, y_int+1)
        value1 = self._avg(value00, value10, x_frac)
        value2 = self._avg(value01, value11, x_frac)
        value  = self._avg(value1,  value2, y_frac)
        #print "%4d %4d | %4d\n%4d %4d | %4d\n-------------\n%4d" % (value00, value10, value1, value01, value11, value2, value)
        return value


#DEBUG ONLY
if __name__ == '__main__':
    x = SRTMDownloader()
    x.loadFileList()
    lat = 49.2
    lon = 11.5
    tile = x.getTile(lat, lon)
    print tile.getAltitudeFromLatLon(lat, lon) #529.1
    print tile.getAltitudeFromLatLon(49.37592, 11.44221) #should be 422m
    #x.createFileList()

File: applications/utils/import/srtm2wayinfo/srtm.cpp
/* Copyright (c) 2009 Hermann Kraus
 * This software is available under a "MIT Style" license
 * (see COPYING).
 */
/** \file
  * SRTM data downloader and tile handler.
  */
#include "srtm.h"
#include "zip.h"

#include <math.h>
#include <QDir>
#include <QStringList>
#include <QString>
#include <QProcess>
#include <QDebug>
#include <qendian.h>

/** Standard error tile that is returned when something goes wrong. */
SrtmTile errorTile("error", -1000, -1000); //TODO

/** Callback from curl for data that is not stored in a file but kept in memory. */
size_t curl_data_callback(void *ptr, size_t size, size_t nmemb, void *stream)
{
    SrtmDownloader *downloader = static_cast<SrtmDownloader*>(stream);
    downloader->curlAddData(ptr, size*nmemb);
    return size*nmemb;
}

/** Callback from curl for data that is stored in a file. */
size_t curl_file_callback(char *ptr, size_t size, size_t nmemb, void *stream)
{
    QFile *file = static_cast<QFile *>(stream);
    int size_left = size * nmemb;
    while (size_left > 0) {
        int result = file->write(ptr, size_left);
        if (result == -1) {
            qCritical() << "Error while writing to file!" << file->errorString();
        } else {
            ptr += result;
            size_left -= result;
        }
    }
    return size*nmemb;
}

/** Constructor for SrtmDownloader.
  * \note This downloader currently only supports SRTM3 data.
  * \param url URL to the SRTM data. 
  * \param cachedir Directory in which the downloaded files are stored.
  */
SrtmDownloader::SrtmDownloader(QString url, QString cachedir)
{
    this->url = url;
    this->cachedir = cachedir+"/";
    regex.setPattern("<a href=\"([NS])(\\d{2})([EW])(\\d{3})\\.hgt\\.zip");
    QDir dir;
    if (!dir.exists(cachedir)) {
        dir.mkpath(cachedir);
    }
    curl = curl_easy_init();
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1);
    curl_easy_setopt(curl, CURLOPT_FAILONERROR, 1); //Make sure error reporting works
    loadFileList();
}

/** One line helper function. */
void SrtmDownloader::curlAddData(void *ptr, int size)
{
    curlData += QString::fromAscii(static_cast<char*>(ptr), size);
}

/** Create a new file list by getting directory contents from server. */
void SrtmDownloader::createFileList()
{
    //Store data in memory
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, curl_data_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, this);
    
    QStringList continents;
    continents << "Africa" << "Australia" << "Eurasia" << "Islands" << "North_America" << "South_America";
    foreach (QString continent, continents) {
        qDebug() << "Downloading data from" << url+continent+"/";
        curlData.clear();
        curl_easy_setopt(curl, CURLOPT_URL, QString(url+continent+"/").toAscii().constData());
        CURLcode error = curl_easy_perform(curl);
        if (error) {
            qCritical() << "Error downloading data for" << continent << "(" << curl_easy_strerror(error) << ")";
        }
        int index = -1;
        while ((index = curlData.indexOf(regex, index+1)) != -1) {
            int lat = regex.cap(2).toInt();
            int lon = regex.cap(4).toInt();
            if (regex.cap(1) == "S") {
                lat = -lat;
            }
            if (regex.cap(3) == "W") {
                lon = - lon;
            }
            //S00E000.hgt.zip
            //123456789012345 => 15 bytes long
            fileList[latLonToIndex(lat, lon)] = continent+"/"+regex.cap().right(15);
        }
    }
    curlData.clear(); //Free mem

    if (fileList.size() != SRTM_FILE_COUNT) {
        qCritical() << "Could not download complete list of tiles from SRTM server. Got" << fileList.size() << "tiles but" << SRTM_FILE_COUNT << "were expected.";
        exit(1);
    }
    
    QFile file(cachedir+"filelist");
    if (!file.open(QIODevice::WriteOnly)) {
        qCritical() << "Could not open file" << cachedir+"filelist";
        //Not a fatal error. We just can't cache the list.
        return;
    }
    QDataStream stream(&file);
    stream << fileList;
    file.close();
}

/** Load a file list or create a new one if it doesn't exist. */
void SrtmDownloader::loadFileList()
{
    QFile file(cachedir+"filelist");
    if (!file.open(QIODevice::ReadOnly)) {
        createFileList();
        return;
    }
    QDataStream stream(&file);
    stream >> fileList;
    file.close();
    if (fileList.size() != SRTM_FILE_COUNT) {
        createFileList();
    }
}

/** Get tile for a specified location.
  * \note The tile object returned is owned by this SrtmDownloader instance. It
  *       _must_ _not_ be deleted by the user. However it _may_ be deleted by
  *       SrtmDownloader during any later call to getTile()
  */
SrtmTile *SrtmDownloader::getTile(float lat, float lon)
{
    int intlat = int(floor(lat)), intlon = int(floor(lon));
    int index = latLonToIndex(intlat, intlon);
    SrtmTile *tile = tileCache[index];

    if (tile) return tile;

    if (fileList.contains(index)) {
        QStringList splitted = fileList[index].split("/", QString::SkipEmptyParts);
        if (!QFile(cachedir + splitted[1]).exists()) {
            downloadTile(fileList[index]);
        }
        tile = new SrtmTile(cachedir + splitted[1], intlat, intlon);
        tileCache.insert(index, tile);
        Q_ASSERT(tileCache[index] != 0);
        return tile;
    } else {
       return &errorTile;
    }
}

/** Download a tile from the server.
    \param filename must be in the format continent/tilename.hgt.zip
    \note You should _not_ call this function when you need tile data. Use getTile instead. */
void SrtmDownloader::downloadTile(QString filename)
{
    qDebug() << "Downloading" << filename;
    QStringList splitted = filename.split("/", QString::SkipEmptyParts);
    
    QFile file(cachedir+splitted[1]);
    if (!file.open(QIODevice::WriteOnly)) {
        qCritical() << "Could not create file" << cachedir+splitted[1];
        return;
    }
    curlData.clear();
    curl_easy_setopt(curl, CURLOPT_URL, QString(url+filename).toAscii().constData());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, curl_file_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &file);
    CURLcode error;
    if ((error = curl_easy_perform(curl))) {
        qCritical() << "Could not download" << filename << "("<< curl_easy_strerror(error) << ")";
    }
    //File is closed automatically
}

/** Get altitude and download necessary tiles automatically. */
float SrtmDownloader::getAltitudeFromLatLon(float lat, float lon)
{
    SrtmTile *tile = getTile(lat, lon);
    if (tile)
        return tile->getAltitudeFromLatLon(lat, lon);
    else
        return SRTM_DATA_VOID;
}

/** Create a new tile object. Unzips the tile data if necessary.
  * \note Special filename: "error" returns an invalid tile.
  */
SrtmTile::SrtmTile(QString filename, int lat, int lon)
{
    this->lat = lat;
    this->lon = lon;
    valid = false;
    if (filename == "error") return;
    buffer = 0;
    size = SrtmZipFile::getData(filename, &buffer);
    Q_ASSERT(size == 1201 || size == 3601);
    Q_ASSERT(buffer != 0);
    valid = true;
}

/** Free resources. */
SrtmTile::~SrtmTile()
{
    if (buffer) delete buffer;
}

/** Get the value of a pixel from the data using a coordinate system
  * starting in the upper left (NW) edge growing to the lower right
  * egde (SE) instead of the SRTM coordinate system.
  */
int SrtmTile::getPixelValue(int x, int y)
{
    Q_ASSERT(x >= 0 && x < size && y >= 0 && y < size);
    int offset = x + size * (size - y - 1);
    qint16 value;
    value = qFromBigEndian(buffer[offset]);
    return value;
}

/** Gets the altitude in meters for a given coordinate. */
float SrtmTile::getAltitudeFromLatLon(float lat, float lon)
{
    if (!valid) return SRTM_DATA_VOID;
    lat -= this->lat;
    lon -= this->lon;
    Q_ASSERT(lat >= 0.0 && lat < 1.0 && lon >= 0.0 && lon < 1.0);
    float x = lon * (size - 1);
    float y = lat * (size - 1);
    /* Variable names:
        valueXY with X,Y as offset from calculated value, _ for average
    */
    float value00 = getPixelValue(x, y);
    float value10 = getPixelValue(x+1, y);
    float value01 = getPixelValue(x, y+1);
    float value11 = getPixelValue(x+1, y+1);
    float value_0 = avg(value00, value10, x-int(x));
    float value_1 = avg(value01, value11, x-int(x));
    float value__ = avg(value_0, value_1, y-int(y));
    return value__;
}